bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -648,6 +651,59 @@ std::pair<PrePrepareMsg *, bool> ReplicaImp::buildPrePrepareMsgBatchByRequestsNu
   return buildPrePrepareMessageByRequestsNum(requiredRequestsNum);
 }
 
+bool ReplicaImp::tryToSendDataPrePrepareMsg() {
+  // no need to check for pre-requsites
+  removeDuplicatedRequestsFromRequestsQueue();
+  PrePrepareMsg *pp = nullptr;
+  bool isSent = false;
+
+  auto batchedReq = reqBatchingLogic_.batchRequests();
+  isSent = batchedReq.second;
+  if (isSent) {
+    pp = batchedReq.first;
+    batch_closed_on_logic_on_++;
+    accumulating_batch_time_.add(
+        std::chrono::duration_cast<std::chrono::microseconds>(getMonotonicTime() - time_to_collect_batch_).count());
+    accumulating_batch_avg_time_.Get().Set((uint64_t)accumulating_batch_time_.avg());
+    if (accumulating_batch_time_.numOfElements() == 1000) {
+      accumulating_batch_time_.reset();  // We reset the average on every 1000 samples
+    }
+    time_to_collect_batch_ = MinTime;
+  }
+
+  if (!pp) return isSent;
+  // check for data separation
+  if (pp->requestsSize() >= config_.threshBatchSizeForDataSeparation) {
+    auto new_pp = getConsensusPPFromDataPP(pp);
+    if (new_pp) {
```

Comment:
Why are we trying to send consensusPP here?
It is likely to assume that we already run consensuses up to the concurrency level threshold, in that case, it is ok only to push an internal message that says: go the the queue and create consensusPP for the first datPP there

Reply:
I mean, we don't need to bind a specific dataPP to the internal message...